File: clients/js/src/requests/player_names.ts
import { BaseRequest } from './base';

export type PlayerNamesRequest = BaseRequest<PlayerNamesRequestType, PlayerNamesRequestAttributes>;

enum PlayerNamesRequestType {
    player_names = 'player_names',
}

export interface PlayerNamesRequestAttributes {
    player: Array<string>;
    rival: Array<string>;
}

export const PlayerNamesRequestDefault = (attributes: PlayerNamesRequestAttributes): PlayerNamesRequest => {
    return {
        data: {
            type: PlayerNamesRequestType.player_names,
            attributes,
        },
    };
};

File: pkmnapi-api/src/requests/player_names.rs
use rocket_okapi::JsonSchema;
use serde::Deserialize;

use crate::requests::base::BaseRequest;

pub type PlayerNamesRequest = BaseRequest<PlayerNamesRequestType, PlayerNamesRequestAttributes>;

impl PlayerNamesRequest {
    pub fn get_player_names(&self) -> &Vec<String> {
        &self.data.attributes.player
    }

    pub fn get_rival_names(&self) -> &Vec<String> {
        &self.data.attributes.rival
    }
}

#[derive(Debug, Deserialize, JsonSchema)]
#[allow(non_camel_case_types)]
pub enum PlayerNamesRequestType {
    player_names,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct PlayerNamesRequestAttributes {
    pub player: Vec<String>,
    pub rival: Vec<String>,
}
